packages/flutter/lib/src/material/pickers/input_date_picker.dart

```diff
@@ -0,0 +1,281 @@
+      return widget.errorFormatText ?? 'Invalid format.';
+    } else if (!_isValidAcceptableDate(date)) {
+      // TODO(darrenaustin): localize 'Out of range.'
+      return widget.errorInvalidText ?? 'Out of range.';
+    }
+    return null;
+  }
+
+  void _handleSaved(String text) {
+    if (widget.onDateSaved != null) {
+      final DateTime date = _parseDate(text);
+      if (_isValidAcceptableDate(date)) {
+        _selectedDate = date;
+        _inputText = text;
+        widget.onDateSaved(date);
+      }
+    }
+  }
+
+  void _handleSubmitted(String text) {
+    if (widget.onDateSubmitted != null) {
+      final DateTime date = _parseDate(text);
+      if (_isValidAcceptableDate(date)) {
+        _selectedDate = date;
+        _inputText = text;
+        widget.onDateSubmitted(date);
+      }
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return OrientationBuilder(builder: (BuildContext context, Orientation orientation) {
+      assert(orientation != null);
+
+      return Container(
+        padding: const EdgeInsets.symmetric(horizontal: 24),
+        height: orientation == Orientation.portrait ? _inputPortraitHeight : _inputLandscapeHeight,
+        child: Column(
+          children: <Widget>[
+            const Spacer(),
+            TextFormField(
+              decoration: InputDecoration(
+                border: const UnderlineInputBorder(),
+                filled: true,
+                // TODO(darrenaustin): localize 'mm/dd/yyyy' and 'Enter Date'
+                hintText: widget.fieldHintText ?? 'mm/dd/yyyy',
+                labelText: widget.fieldLabelText ?? 'Enter Date',
+              ),
+              validator: _validateDate,
+              inputFormatters: <TextInputFormatter>[
+                // TODO(darrenaustin): localize date separator '/'
+                _DateTextInputFormatter('/'),
+              ],
+              keyboardType: TextInputType.datetime,
+              onSaved: _handleSaved,
+              onFieldSubmitted: _handleSubmitted,
+              autofocus: widget.autofocus,
+              controller: _controller,
+            ),
+            const Spacer(),
+          ],
+        ),
+      );
+    });
+  }
+}
+
+class _DateTextInputFormatter extends TextInputFormatter {
+
+  _DateTextInputFormatter(this.separator);
+
+  final String separator;
+
+  final WhitelistingTextInputFormatter _filterFormatter =
+    // Only allow digits and separators (slash, dot, comma, hyphen, space).
+    WhitelistingTextInputFormatter(RegExp(r'[\d\/\.,-\s]+'));
+
+  @override
+  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
+    final TextEditingValue filteredValue = _filterFormatter.formatEditUpdate(oldValue, newValue);
+    return filteredValue.copyWith(
+      // Replace any separator character with the given separator
+      text: filteredValue.text.replaceAll(RegExp(r'[\D]'), separator),
+    );
+  }
+}
```

Comment:
@darrenaustin I'm about to check in something that replaces WhitelistingTextInputFormatter and BlacklistingTextInputFormatter with a single FilteringTextInputFormatter that handles replacement even for the case of allow-list filtering, which you might be able to use here.